Rename `SkipLinkController` to `FocusController`, update unit tests & add story

lb- opened the pull request:
## Description

The `SkipLinkController` was one of the first controllers implemented and as such has not adopted some of the practice that's emerged through other Stimulus adoption work.

## Details

This PR introduces the following:

1. A simpler name that reflects the **behaviour** not the **usage** of the Controller
2. Better unit tests with a structure that's aligned with many other unit testing
3. Better JSDoc leverage with proper `@example` tags
4. Abstract the common add/remove event listener behaviour and only remove the tabindex if the target element did not have one in the first place - pull this out to a robust util `forceFocus` as we use this in lots of other place in the code
5. Enhance the controller to be able to trigger focus on any targeted element, so we do not have to use it for links only, also ensure we can smooth scroll into view before triggering the focus for a better user experience
6. Add a Story that covers the example usage for additional manual testing
7. Add a Python unit test that covers the presence of the skip link, this is not every page but at least ensures we do not accidentally remove this code from the base admin template
8. Fix up an issue with our stubs/adapter files so we can better test scrolling with Jest function mocks

See also https://github.com/wagtail/wagtail/pull/12554 (a similar, in purpose, rename & clean up effort).

## How to test

Within the admin interface, once refreshed, press `tab`. The first target element should be the 'skip to main content' link, activating this (press enter) should move your focus away and into the main content. Pressing `tab` again, should focus on the first focusable element in the main content (e.g. the first link '34 pages' in the screen below).

<img width="1800" alt="Screenshot 2024-11-08 at 7 53 46 AM" src="https://github.com/user-attachments/assets/3dcf7944-cf92-498b-97e7-ff3919454556">

## Why this matters

Eventually we want to consider broader usage of the focus controller, either made available through documentation or for 'scroll to top' type buttons around longer page content. This is already a request for a different issue - https://github.com/wagtail/wagtail/issues/12128

This also makes it easier for us to eventually centralise the Jest testing for Stimulus controllers, the `await setup` pattern is super useful and makes adding future unit tests much easier.

Finally, the fact that the element's name (Link) is in the controller is a bit of an anti-pattern, any element could be used (e.g. a button) and should normally not be reflected in the controller name.

While renames seem unnecessary, without this rename it may seem like we need a different controller do do any 'focus' work, when in fact Stimulus shines when we create smaller modular 'behaviour' controllers and not usage specific ones. This aligns with our Stimulus tips, which were not really full formed when this controller was created. https://docs.wagtail.org/en/stable/contributing/ui_guidelines.html#helpful-tips

laymonage reviewed (commented):
Could you remind me why this had to use a Stimulus controller over just using the `<a href` tag? 🤔

lb- commented:
It's just a migration from the original JS that was there (links below).

My understanding is that while a plain link can work OK, it's a better experience with the focus outline being triggered. In all the sites I've worked on, there's always been a JS behaviour attached (link or buttons vary).

For context:
- https://github.com/wagtail/wagtail/pull/9818
- https://github.com/wagtail/wagtail/pull/5780

More than happy for us to remove the JS completely but we'd probably have to consider other implications of this and get some accessibility team input first.

lb- commented:
Just updated the comment above.

lb- commented:
This PR would be great to get in as it will help us build on it for supporting https://github.com/wagtail/wagtail/issues/12128 & https://github.com/wagtail/wagtail/pull/12739

lb- commented:
I have updated this PR to pull out the `forceFocus` util to a utils file, added extra handling for scrolling, async delay for 'just created' elements and added unit tests for this. This way it can be used as is for future work on https://github.com/wagtail/wagtail/issues/12128 & https://github.com/wagtail/wagtail/pull/12739 @Srishti-j18

In addition, I have fixed up some issues with the `stubs`/`adapter` and added clearer docs at the top of the file, we were including items in subs that should have been unit tests only. I will try to create this small fix as a stand alone PR also.

lb- commented:
CI failure is unrelated to this PR.

```
FAIL: test_ordering_by_content_type (wagtail.admin.tests.pages.test_explorer_view.TestPageExplorer) (ordering='-content_type')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/wagtail/wagtail/wagtail/admin/tests/pages/test_explorer_view.py", line 192, in test_ordering_by_content_type
    self.assertEqual(page_ids, pages)
AssertionError: Lists differ: [1094, 1095, 1093] != [1094, 1093, 1095]

First differing element 1:
1095
1093
```

lb- commented on client/src/controllers/FocusController.ts:
```diff
@@ -0,0 +1,56 @@
+    const selector =
+      this.targetValue || this.element.getAttribute('href') || 'main';
+
+    this.target =
```
Note. We should move this logic to the focus method. And add a unit test for the focus target appearing after initial render.

lb- replied:
Might be nice to add a unit test for an element created at the same time as the focus is clicked. This is why we want to defer by a delay. Maybe using the clone controller in the unit test.

lb- commented on client/src/utils/forceFocus.test.js:
```diff
@@ -0,0 +1,44 @@
+    button.focus();
+    expect(document.activeElement).toBe(button);
+
+    expect(element.getAttribute('tabindex')).toBe(null);
```
Add another test for an element 'just created'.

Srishti-j18 commented:
Should these changes be made in this PR or in https://github.com/wagtail/wagtail/pull/12739? I'm a little confused since I merged all the commits from this PR into https://github.com/wagtail/wagtail/pull/12739.

lb- commented:
@Srishti-j18  I have updated this PR to include a better approach to the `focus` & `target` handling so it allows for focusing on elements that did not exist when the controller connected.

Feel free to incorporate this however you need to get your PR working reliably, we can clean up your PR later once this PR is in. I have asked for a review on this within the core team but I am not sure when that will happen.

lb- commented:
OK - this is now rebased on top of the latest main (which includes the CI fix + stubs/adapter clean up)

laymonage commented on wagtail/admin/tests/tests.py:
```diff
@@ -31,6 +31,15 @@ def test_simple(self):
+    def test_accessible_skip_link_exists(self):
+        response = self.client.get(reverse("wagtailadmin_home"))
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(
+            response,
+            '<a class="skiplink button" href="#main" data-controller="w-focus" data-action="w-focus#focus:prevent">Skip to main content</a>',
+        )
+        self.assertContains(response, ' id="main">')
```
Would it be better to use BeautifulSoup here so we don't have to assert `class="skiplink button"` and can actually select an element with `id="main"`?

lb- replied:
Good idea, done.

laymonage reviewed (commented):
Thank you! Just some minor comments but I'm very happy with the refactoring here.

lb- commented on client/src/utils/forceFocus.ts:
```diff
@@ -0,0 +1,41 @@
+
+    if (element instanceof HTMLElement) {
+      element.focus({ preventScroll });
+    }
```
Thanks for this, I think you are right, the current code kind of mixed responsibilities, both being defensive.

I have opted for the `forceFocus` to NOT accept `null`, but also for completeness allowed `SVGElement` which also allows `focus()`.

I then changed the signature for the `FocusController` selectors. If somehow an `Element` gets selected, if that's even possible for that code, it will throw an error. Which is what we would expect. The error will throw inside the promise, but I think that will be enough handling for now.

lb- commented:
Thanks @laymonage
